lib/logstash/outputs/loggly.rb

```diff
@@ -88,31 +105,114 @@ class LogStash::Outputs::Loggly < LogStash::Outputs::Base
 
   public
   def register
-    # nothing to do
   end
 
   public
+  def multi_receive(events)
+    send_batch events.collect { |event| prepare_meta(event) }
+  end
+
   def receive(event)
+    send_batch [prepare_meta(event)]
+  end
+
+  private
+  # Returns one meta event {key: '...', tag: '...', event: event },
+  # or returns nil, if event's key doesn't resolve.
+  def prepare_meta(event)
     key = event.sprintf(@key)
     tag = event.sprintf(@tag)
 
+    if expected_field = key[/%{(.*)}/, 1]
+      @logger.warn "Skipping sending message to Loggly. No key provided (key='#{key}'). Make sure to set field '#{expected_field}'."
+      @logger.debug "Dropped message", :event => event.to_json
+      return nil
+    end
+
     # For those cases where %{somefield} doesn't exist
     # we should ship logs with the default tag value.
-    tag = 'logstash' if /^%{\w+}/.match(tag)
+    tag = DEFAULT_LOGGLY_TAG if /%{\w+}/.match(tag)
 
-    # Send event
-    send_event("#{@proto}://#{@host}/inputs/#{key}/tag/#{tag}", format_message(event))
-  end # def receive
+    meta_event = {  key: key, tag: tag, event: event }
```

Comment:
can we use 1.9 style hashes here? this guarantees it will work in logstash 5.6

Reply:
Yes, these were introduced in Ruby 1.9. Do you see an issue with using this style of hashes?